db/core.py

```diff
@@ -27,4 +34,154 @@ class BaseModel(SQLModel):
+        return self
+
+    @classmethod
+    def get_deleted_by_id(cls, session: Session, identity: Any) -> "SoftDeleteModel | None":
+        """
+        Retrieve a soft-deleted record by primary key.
+        """
+        identity_dict = cls._coerce_primary_key_map(identity)
+        stmt = (
+            select(cls)
+            .execution_options(include_deleted=True)
+            .filter_by(**identity_dict)
+            .where(cls.is_deleted.is_(True))
+        )
+        return session.exec(stmt).scalars().one_or_none()
+
+    @classmethod
+    def _coerce_primary_key_map(cls, identity: Any) -> dict[str, Any]:
+        """
+        Coerce arbitrary primary key identifiers into a ``{column: value}`` mapping.
+
+        ``identity`` may be provided as:
+        * a dict where keys match the primary key columns,
+        * a single scalar value when the model uses a single-column primary key,
+        * a tuple/list containing values for each primary key column in order.
+
+        Any mismatch between the provided structure and the model's primary key
+        definition raises ``ValueError`` so downstream queries remain predictable.
+        """
+        mapper = sa_inspect(cls)
+        pk_cols = mapper.primary_key
+        if not pk_cols:
+            raise ValueError(f"{cls.__name__} does not have a primary key defined.")
+
+        if isinstance(identity, dict):
+            return identity
+
+        if len(pk_cols) == 1 and not isinstance(identity, (tuple, list)):
+            return {pk_cols[0].key: identity}
+
+        if isinstance(identity, (tuple, list)):
+            if len(identity) != len(pk_cols):
+                raise ValueError(
+                    f"Identity length {len(identity)} does not match primary key length {len(pk_cols)}"
+                )
+            return {col.key: value for col, value in zip(pk_cols, identity)}
+
+        raise ValueError("Identity must be scalar, tuple/list, or dict matching the primary key.")
+
+
+def _copy_column_state(source: SoftDeleteModel, target: SoftDeleteModel) -> None:
+    mapper = sa_inspect(source.__class__)
+    for attr in mapper.column_attrs:
+        key = attr.key
+        if key == "is_deleted":
+            continue
+        setattr(target, key, getattr(source, key))
+
+
+def _identity_dict_from_instance(instance: SoftDeleteModel) -> dict[str, Any] | None:
+    mapper = sa_inspect(instance.__class__)
+    identity: dict[str, Any] = {}
+    for column in mapper.primary_key:
+        value = getattr(instance, column.key, None)
+        if value is None:
+            return None
+        identity[column.key] = value
+    return identity
+
+
+def _soft_delete_filter(cls) -> Any:
+    mapper = sa_inspect(cls, raiseerr=False)
+    if mapper is None:
+        return expression.true()
+    column = mapper.c.is_deleted
+    return column.is_(False)
+
+
+@event.listens_for(SASession, "before_flush")
+def _revive_soft_deleted(session: SASession, flush_context, instances) -> None:
```

Comment:
When does this trigger? When attempting to create a new object with the same primary key as a deleted one?